vendors/nuvoton/sdk/middleware/wifi_esp8266/esp8266_wifi.c

```diff
@@ -364,16 +453,41 @@ ESP_WIFI_Status_t ESP_IO_Recv( ESP_WIFI_Object_t * pxObj, uint8_t pucRxBuf[], ui
                     pcPtr = strtok(NULL, pcDelim);
                     ucIpdLinkID = (uint8_t)atoi(pcPtr);
                 }
+                /* Get the IPD length */
                 pcPtr = strtok(NULL, pcDelim);
                 usIpdLength = (uint16_t)atoi(pcPtr);
-                if (xWifiIpd[ucIpdLinkID].DataLength + usIpdLength > sizeof(xWifiIpd[ucIpdLinkID].Data)) {
-                    if (xWifiIpd[ucIpdLinkID].DataLength < sizeof(xWifiIpd[ucIpdLinkID].Data)) {
-                        configPRINTF(("ERROR: [%s] Get ipd %d bytes reach the maximum size !!\n", __func__, usIpdLength));
-                    }
+
+                if (ucIpdLinkID >= wificonfigMAX_SOCKETS || !xWifiIpd[ucIpdLinkID].IsOpen) {
```

Comment:
I have been running some unit tests and one is consistency failing with the message in this block. That did not appear before in the Master branch.

```
8 156346 [NetRecv] ERROR: [ESP_WIFI_Recv] Invalid socket ID 1 !!
17:46:06 
TEST(MQTT_System, LastWillAndTestament)../../../../../libraries/c_sdk/standard/mqtt/test/system/iot_tests_mqtt_system.c:953::FAIL: Expected 0 Was 5
```

Also the WiFi tests which pass on Master consistently fail on this branch.

Could you please run them locally:

https://github.com/aws/amazon-freertos/blob/master/vendors/nuvoton/boards/numaker_iot_m487_wifi/aws_tests/config_files/aws_test_runner_config.h#L55

Reply:
@sarenameas,  After cross checking, it could pass the previous version of wifi test code but fail in the current version of test code. At least, the old wifi driver without flow control could pass the current test code.
So, it's not too difficult and we will fix it.

Reply:
@sarenameas , We fixed the WiFi unit test failure by commit 381e87d, the last update will not block CPU to avoid unit-test failure.

Reply:
@cyliangtw, some of the `Full_WiFi` unit tests are still failing. Following are the test results. Can you please take a look at them and make sure that they succeed?
```
{
    "Full_WiFi": {
        "AFQP_WIFI_Scan": {
            "status": "PASS"
        },
        "AFQP_WIFI_GetMode_NullParameters": {
            "status": "PASS"
        },
        "AFQP_WIFI_GetIP_NullParameters": {
            "status": "FAIL",
            "details": "../../../../../libraries/abstractions/wifi/test/iot_test_wifi.c:938::FAIL: Expression Evaluated To FALSE"
        },
        "AFQP_WIFI_GetMAC_NullParameters": {
            "status": "FAIL",
            "details": "../../../../../libraries/abstractions/wifi/test/iot_test_wifi.c:991::FAIL: Expression Evaluated To FALSE"
        },
        "AFQP_WIFI_GetHostIP_NullParameters": {
            "status": "FAIL",
            "details": "../../../../../libraries/abstractions/wifi/test/iot_test_wifi.c:1037::FAIL: Expression Evaluated To FALSE"
        },
        "AFQP_WIFI_Scan_NullParameters": {
            "status": "PASS"
        },
        "AFQP_WIFI_NetworkAdd_NullParameters": {
            "status": "PASS"
        },
        "AFQP_WIFI_NetworkGet_NullParameters": {
            "status": "PASS"
        },
        "AFQP_WIFI_SetPMMode_NullParameters": {
            "status": "PASS"
        },
        "AFQP_WIFI_GetPMMode_NullParameters": {
            "status": "PASS"
        },
        "AFQP_WIFI_Ping_NullParameters": {
            "status": "FAIL",
            "details": "../../../../../libraries/abstractions/wifi/test/iot_test_wifi.c:1700::FAIL: Expression Evaluated To FALSE"
        },
        "AFQP_WIFI_ConnectAP_NullParameters": {
            "status": "PASS"
        },
        "AFQP_WIFI_SetMode_InvalidMode": {
            "status": "PASS"
        },
        "AFQP_WIFI_GetHostIP_InvalidDomainName": {
            "status": "FAIL",
            "details": "../../../../../libraries/abstractions/wifi/test/iot_test_wifi.c:1063::FAIL: Expression Evaluated To FALSE"
        },
        "AFQP_WIFI_GetHostIP_DomainNameLengthExceeded": {
            "status": "FAIL",
            "details": "../../../../../libraries/abstractions/wifi/test/iot_test_wifi.c:1089::FAIL: Expression Evaluated To FALSE"
        },
        "AFQP_WIFI_NetworkDelete_DeleteNonExistingNetwork": {
            "status": "PASS"
        },
        "AFQP_WIFI_NetworkGetNonExistingNetwork": {
            "status": "PASS"
        },
        "AFQP_WIFI_SetPMMode_InvalidPMMode": {
            "status": "PASS"
        },
        "AFQP_WIFI_ConnectAP_InvalidSSID": {
            "status": "PASS"
        },
        "AFQP_WIFI_ConnectAP_InvalidPassword": {
            "status": "PASS"
        },
        "AFQP_WIFI_ConnectAP_MaxSSIDLengthExceeded": {
            "status": "PASS"
        },
        "AFQP_WIFI_ConnectAP_MaxPasswordLengthExceeded": {
            "status": "PASS"
        }
    }
}
```

Reply:
@aggarw13 ,
      My test log show it could pass all of your wifi unit tests. I guess your M487 board on hand maybe older version. Mine is NuMaker-IoT--M487 **v1.3**, if yours is v1.2 or previous version, the AT firmware of wifi module is too old and maybe unstable in some connection use case.

**My test log**:
0 0 [None] FreeRTOS_IPInit
1 0 [None] Write certificate...
2 0 [None] vTaskStartScheduler
3 318 [Tmr Svc] Wi-Fi module initialized. Connecting to AP...
4 318 [Tmr Svc] current task name is "Tmr Svc"
5 625 [Tmr Svc] Try to connect to AP "HUAWEI-F2C9" ...
6 5703 [Tmr Svc] Wi-Fi Connected to AP. Creating tasks which use network...
7 5703 [Tmr Svc] IP Address acquired 192.168.8.106
---------STARTING TESTS---------
8 8653 [TestRunner] WiFi Networks and strength:
9 8653 [TestRunner]     NTCWLAN-N: -32
10 8653 [TestRunner]     NTGuest: -33
11 8653 [TestRunner]     WECWLAN_11n: -33
12 8653 [TestRunner]     git_server: -43
13 8653 [TestRunner]     NTGuest: -58
14 8653 [TestRunner]     NTGuest: -75
15 8653 [TestRunner]     WECWLAN_11n: -58
16 8653 [TestRunner]     WECWLAN_11n: -74
17 8653 [TestRunner]     likelike: -46
18 8655 [TestRunner]     WECWLAN_11n: -40
19 8655 [TestRunner] End of WiFi Networks
TEST(Full_WiFi, AFQP_WIFI_Scan) PASS
TEST(Full_WiFi, AFQP_WIFI_GetMode_NullParameters) PASS
20 10879 [TestRunner] current task name is "TestRunner"
21 11186 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
TEST(Full_WiFi, AFQP_WIFI_GetIP_NullParameters) PASS
22 18376 [TestRunner] current task name is "TestRunner"
23 18683 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
TEST(Full_WiFi, AFQP_WIFI_GetMAC_NullParameters) PASS
24 31873 [TestRunner] current task name is "TestRunner"
25 32179 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
TEST(Full_WiFi, AFQP_WIFI_GetHostIP_NullParameters) PASS
TEST(Full_WiFi, AFQP_WIFI_Scan_NullParameters) PASS
TEST(Full_WiFi, AFQP_WIFI_NetworkAdd_NullParameters) PASS
TEST(Full_WiFi, AFQP_WIFI_NetworkGet_NullParameters) PASS
TEST(Full_WiFi, AFQP_WIFI_SetPMMode_NullParameters) PASS
TEST(Full_WiFi, AFQP_WIFI_GetPMMode_NullParameters) PASS
26 39930 [TestRunner] current task name is "TestRunner"
27 40236 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
TEST(Full_WiFi, AFQP_WIFI_Ping_NullParameters) PASS
28 47427 [TestRunner] current task name is "TestRunner"
29 47733 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
30 62892 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
31 77261 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
TEST(Full_WiFi, AFQP_WIFI_ConnectAP_NullParameters) PASS
TEST(Full_WiFi, AFQP_WIFI_SetMode_InvalidMode) PASS
32 91551 [TestRunner] current task name is "TestRunner"
33 91857 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
TEST(Full_WiFi, AFQP_WIFI_GetHostIP_InvalidDomainName) PASS
34 112994 [TestRunner] current task name is "TestRunner"
35 113301 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
TEST(Full_WiFi, AFQP_WIFI_GetHostIP_DomainNameLengthExceeded) PASS
TEST(Full_WiFi, AFQP_WIFI_NetworkDelete_DeleteNonExistingNetwork) PASS
TEST(Full_WiFi, AFQP_WIFI_NetworkGetNonExistingNetwork) PASS
TEST(Full_WiFi, AFQP_WIFI_SetPMMode_InvalidPMMode) PASS
36 134710 [TestRunner] current task name is "TestRunner"
37 135016 [TestRunner] Try to connect to AP "invalid ssid" ...
38 149389 [TestRunner] Try to connect to AP "invalid ssid" ...
39 163762 [TestRunner] Try to connect to AP "invalid ssid" ...
TEST(Full_WiFi, AFQP_WIFI_ConnectAP_InvalidSSID) PASS
40 177941 [TestRunner] current task name is "TestRunner"
41 178247 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
42 193405 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
43 207776 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
TEST(Full_WiFi, AFQP_WIFI_ConnectAP_InvalidPassword) PASS
44 221953 [TestRunner] current task name is "TestRunner"
45 222259 [TestRunner] Try to connect to AP "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"46 222579 [TestRunner] Try to connect to AP "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"47 222899 [TestRunner] Try to connect to AP "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"TEST(Full_WiFi, AFQP_WIFI_ConnectAP_MaxSSIDLengthExceeded) PASS
48 223026 [TestRunner] current task name is "TestRunner"
49 223332 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
50 238495 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
51 252867 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
TEST(Full_WiFi, AFQP_WIFI_ConnectAP_MaxPasswordLengthExceeded) PASS
52 269376 [TestRunner] current task name is "TestRunner"
53 269682 [TestRunner] Try to connect to AP "HUAWEI-F2C9" ...
54 279550 [TestRunner] Wi-Fi reconnected following tests finished.
TEST(Quarantine_WiFi, AFQP_WiFiOnOff) PASS
TEST(Quarantine_WiFi, AFQP_WiFiMode) PASS